library/Translate/src/Juvix/BerlinPasses/Contextify.hs

```diff
@@ -0,0 +1,347 @@
+
+unknownSymbolLookupTransform ::
+  ( HasThrow "error" Sexp.T m,
+    Feedback.Eff m,
+    Env.HasClosure m,
+    MonadIO m
+  ) =>
+  Context.T ->
+  Sexp.T ->
+  m Sexp.T
+unknownSymbolLookupTransform ctx =
+  Passes.primiveOrSymbol ctx |> traverseOnDeserialized
+
+--------------------------------------------------------------------------------
+-- Helpers for Constructing Passes
+--------------------------------------------------------------------------------
+
+mkTrans ::
+  NameSymbol.T ->
+  -- | The name of the transform to be traced
+  ( Context.T ->
+    Sexp.T ->
+    Env.MinimalMIO Sexp.T
+  ) ->
+  Automation.SimplifiedPassArgument ->
+  Env.MinimalMIO Automation.Job
+mkTrans name trans simplify =
+  Trace.withScope scopeName [show (simplify ^. current)] $
+    trans (simplify ^. context) (simplify ^. current)
+      >>| (`Automation.ProcessNoEnv` [])
+      >>| Automation.ProcessJob
+  where
+    scopeName = "Context" <> name <> "trans"
+
+mkPass ::
+  NameSymbol.T ->
+  -- | The name of the transform to be traced
+  ( Automation.SimplifiedPassArgument ->
+    Env.MinimalMIO Automation.Job
+  ) ->
+  Step.Named
+mkPass name trans =
+  ( Trace.withScope
+      runnerName
+      []
+      . simplifiedTrans
+  )
+    |> Automation.runSimplifiedPass
+    |> Step.T
+    |> Step.namePass passName
+  where
+    simplifiedTrans = Automation.simplify trans
+    runnerName = passName <> "runner"
+    passName = "Context" <> name
+
+--------------------------------------------------------------------------------
+-- Record Recognition Transformation
+--------------------------------------------------------------------------------
+
+recordDeclarationPass :: Step.Named
+recordDeclarationPass =
+  ( Trace.withScope
+      "Context.recordDeclaration-runner"
+      []
+      . recordDeclaration
+  )
+    |> Automation.runSimplifiedPass
+    |> Step.T
+    |> Step.namePass "Context.recordDeclaration"
+
+recordDeclaration ::
+  Pipeline.PassArgument ->
+  Env.MinimalMIO Pipeline.Job
+recordDeclaration pa =
+  case pa ^. current of
+    Pipeline.InContext name ->
+      case ctx Context.!? name of
+        Just from -> do
+          let qualifiedName = from ^. Context.qualifedName
+          newRecord <-
+            from ^. Context.term
+              |> Passes.figureRecord
+              >>| Context.formPutBack
+          updateJob qualifiedName newRecord |> pure
+        Nothing ->
+          Juvix.Library.throw @"error" $
+            Sexp.string
+              "Could not find definition when \
+              \ it was promised to be in the environment"
+    _ -> noOpJob |> pure
+  where
+    ctx = pa ^. context
+    noOpJob = Automation.noOpJob ctx (pa ^. current)
+    updateJob name info =
+      Pipeline.UpdateJob
+        { newContext = Context.addGlobal name info ctx,
+          process =
+            Pipeline.Process
+              { _current = Pipeline.InContext name,
+                _newForms = []
+              }
+        }
```

Comment:
I think once we merge this, we can make this pass a `simplified`. As it looks like we've made more work for ourselves by just binding over the pass that is one (figureRecord just sets the def and doesn't return any additional forms.)

Reply:
You're right - I'll change this.